pallets/asset-index/src/lib.rs

```diff
@@ -215,19 +235,108 @@ pub mod pallet {
+                    .and_then(|units| price.reciprocal_volume(units))
+                    .ok_or(Error::<T>::AssetVolumeOverflow)
+                    .and_then(|units| {
+                        units.try_into().map_err(|_| Error::<T>::AssetUnitsOverflow)
+                    })?;
+            }
 
-            // TODO if total supply of any asset drops to 0 it gets removed from the index.
+            // update the index balance by burning all of the redeemed tokens and
+            // issuing new tokens for the tx fee into the treasury's account
+            let burned = T::IndexToken::burn(amount);
+            if let Err(burned) = T::IndexToken::settle(
+                &caller,
+                burned,
+                WithdrawReasons::TRANSFER,
+                ExistenceRequirement::KeepAlive,
+            ) {
+                // TODO can this even happen after ensure_can_withdraw?
+                T::IndexToken::issue(burned.peek());
+                return Err(Error::<T>::InsufficientDeposit.into());
+            }
 
-            todo!()
+            let fee = T::IndexToken::issue(fee);
+            T::IndexToken::resolve_creating(&T::TreasuryPalletId::get().into_account(), fee);
+
+            let mut assets = Vec::with_capacity(asset_prices.len());
+            // start bonding and locking
+            for (price, units) in asset_prices {
+                let asset = price.quote;
+                // try to start the unbonding process
+                let state = if T::RemoteAssetManager::unbond(asset.clone(), units).is_ok() {
+                    // the XCM call was dispatched successfully, however, this is
+                    //  *NOT* synonymous with a successful completion of the unbonding process.
+                    //  instead, this state implies that XCM is now being processed on a different parachain
+                    RedemptionState::Unbonding
+                } else {
+                    // the manager encountered an error before being able to send the XCM call,
+                    //  nothing was dispatched to another parachain
+                    RedemptionState::Initiated
+                };
+
+                // update the holding balance
+                Holdings::<T>::mutate_exists(&asset, |maybe_asset_data| {
+                    if let Some(mut data) = maybe_asset_data.take() {
+                        data.units = data.units.saturating_sub(units);
+                        if !data.units.is_zero() {
+                            return Some(data);
+                        }
+                    }
+                    None
+                });
+
+                assets.push(AssetWithdrawal {
+                    asset,
+                    state,
+                    units,
+                });
+            }
+
+            // lock the assets for the withdrawal period starting at current block
+            PendingWithdrawals::<T>::mutate(&caller, |maybe_redemption| {
```

Comment:
Maybe we can add docs to our mutating operations to declare our mutations are safe when our function doesn't have a wrapper like `with_transaction_result`